otter/convergence/service.py

```diff
@@ -56,99 +62,363 @@ def all_met(server, current_lb_nodes):
+    if failure.check(ConcurrentError):
+        # We don't need to spam the logs about this, it's to be expected
+        pass
+    # TODO: Check for other fatal errors, and put the group into ERROR
+    # state.
+    else:
+        log.err(failure, 'converge-non-fatal-error')
+        # We specifically don't clean up the dirty flag in this case, so
+        # convergence will be retried.
+
+
+def _stable_hash(s):
+    """Get a stable hash of a string as an integer."""
+    # :func:`hash` is not stable with different pythons/architectures.
+    return int(sha1(s).hexdigest(), 16)
+
+
+class Converger(MultiService):
+    """
+    A service that searches for groups that need converging and then does the
+    work of converging them.
+
+    This service is pretty much just a wrapper for :func:`execute_convergence`,
+    with a few important layers above it:
+
+    - virtual "buckets" are partitioned between nodes running this service by
+      using ZooKeeper (thus, this service could/should be run separately from
+      the API). group IDs are deterministically mapped to these buckets.
+    - we repeatedly check for 'dirty flags' created by the
+      :obj:`ConvergenceStarter` service, and determine if they're "ours" with
+      the partitioner.
+    - we ensure we don't execute convergence for the same group concurrently.
+    """
+
+    def __init__(self, log, dispatcher, buckets, partitioner_factory,
+                 converge_all=converge_all):
+        """
+        :param log: a bound log
+        :param dispatcher: The dispatcher to use to perform effects.
+        :param buckets: collection of logical `buckets` which are shared
+            between all Otter nodes running this service. Will be partitioned
+            up between nodes to detirmine which nodes should work on which
+            groups.
+        :param partitioner_factory: Callable of (log, callback) which should
+            create an :obj:`Partitioner` to distribute the buckets.
         """
-        Converge a group to a capacity with a launch config.
-
-        :param log: a bound logger.
-        :param IScalingGroup scaling_group: The scaling group object.
-        :param GroupState group_state: The group state.
-        :param launch_config: An otter launch config.
-        :param perform: :func:`perform` function to use for testing.
-        :param execute_convergence: :func:`execute_convergence` function to use
-            for testing.
+        MultiService.__init__(self)
+        self._dispatcher = dispatcher
+        self._buckets = buckets
+        self.log = log.bind(system='converger')
+        self.partitioner = partitioner_factory(self.log, self.buckets_acquired)
+        self.partitioner.setServiceParent(self)
+        self.group_locks = make_lock_set()
+        self._converge_all = converge_all
+
+    def buckets_acquired(self, my_buckets):
+        """
+        Perform the effectful result of :func:`check_convergence`.
```

Comment:
I am not sure where but somewhere there should be doc mentioning this function is called every 10 seconds. Unfortunately, the 10 is `partial`ed in `partitioner_factory` above. So, changing the seconds from caller will outdate the doc here.